### PR TITLE
Changes for xpcs workflow

### PR DESCRIPTION
**Pull request by birdzix**
See:
https://github.com/Xi-CAM/Xi-cam.SAXS/pull/34
https://github.com/Xi-CAM/Xi-cam.XPCS/pull/53

Adds extra visualization support and data selection (dataselectorview) features (active ensembles) in support of XPCS/SAXS visual refactor.

### CODE REVIEW

File: xicam/gui/models/__init__.py [hunk 1]
```diff
@@ -19,52 +20,116 @@ class EnsembleModel(TreeModel):
...
     canvas_role = Qt.UserRole + 3
+    active_role = Qt.UserRole + 4  # only tied to Ensemble tree items
 
     def __init__(self, parent=None):
         super(EnsembleModel, self).__init__(parent)
-        self.rootItem.setData("Ensembles", Qt.DisplayRole)
+
+        # Keep track of current active item (can be 0 or 1 active items)
+        self.active_ensemble = None
+        self._active_ensemble_name = "(None)"
```
Comment (ronpandolfi) [inline xicam/gui/models/__init__.py]: single source static value

File: xicam/gui/models/__init__.py [hunk 2]
```diff
@@ -19,52 +20,116 @@ class EnsembleModel(TreeModel):
...
+        # Unnamed items (i.e. no display role text) will get this text set
+        self._defaultDisplayText = "Untitled"
```
Comment (ronpandolfi) [inline xicam/gui/models/__init__.py]: make class attr

File: xicam/gui/models/__init__.py [hunk 3]
```diff
@@ -19,52 +20,116 @@ class EnsembleModel(TreeModel):
...
+        # Display the active item as the root title as well
+        self.rootItem.setData(f"Active Item: {self.active_ensemble_name}", Qt.DisplayRole)
+
+    def _update_title(self, name):
+        print(f"\nupdating the title to {name}")
```
Comment (ronpandolfi) [inline xicam/gui/models/__init__.py]: prune print

File: xicam/gui/models/__init__.py [hunk 4]
```diff
@@ -19,52 +20,116 @@ class EnsembleModel(TreeModel):
...
+                print(f"current active item: {self.active_ensemble}")
+                if self.active_ensemble is not None:
+                    # self.active_ensemble.setData(False, self.active_role)
+                    active_index = self.index(self.active_ensemble.row(), 0)
+                    print(f"disabling the previous active ensemble {active_index} ({active_index.data(Qt.DisplayRole)})")
+                    self.setData(active_index, False, self.active_role)
+                print(f"update the active_ensemble to {item} ({item.itemData[Qt.DisplayRole]})")
+                self.active_ensemble = item
+                self._update_title(self.active_ensemble_name)
+                brush = QBrush(Qt.red)
```
Comment (ronpandolfi) [inline xicam/gui/models/__init__.py]: look up qpallete(?)

File: xicam/gui/models/__init__.py [hunk 5]
```diff
@@ -19,52 +20,116 @@ class EnsembleModel(TreeModel):
...
-                logMessage(e, level=WARNING)
-
-            ensemble_item.appendChild(catalog_item)
+            self._create_catalog_item(ensemble_item, catalog, projector)
 
         self.rootItem.appendChild(ensemble_item)
-        self.layoutChanged.emit()
+        # ensemble_item.setData(True, self.active_role)
+        index = self.index(ensemble_item.row(), 0)
+        self.setData(index, True, self.active_role)
```
Comment (ronpandolfi) [inline xicam/gui/models/__init__.py]: add conditional to only do this if its the first(/last) ensemble

File: xicam/gui/models/__init__.py [hunk 6]
```diff
@@ -80,6 +145,40 @@ def rename_ensemble(self, ensemble, name):
...
+    def data(self, index: QModelIndex, role=Qt.DisplayRole) -> Any:
+        item = self.getItem(index)
+
+        if role == self.active_role:
+            if item == self.active_ensemble:
+                return True
+            return False
+
+        # Handle case where display text not provided
+        if role == Qt.DisplayRole:
```
Comment (ronpandolfi) [inline xicam/gui/models/__init__.py]: use elif

File: xicam/gui/models/treemodel.py [hunk 7]
```diff
@@ -1,7 +1,8 @@
 from typing import Any
 from qtpy.QtCore import Qt, QAbstractItemModel, QModelIndex
 
-
+# TODO: refactor check_state (don't need custom methods, use setData)
+# TODO: refactor flags (don't need custom methods? use setData)
```
Comment (ronpandolfi) [inline xicam/gui/models/treemodel.py]: flags are probably fine; lets fix check_state to use roles

File: xicam/gui/plugins/ensembleguiplugin.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
+from typing import Callable
+
+from databroker.core import BlueskyRun
+from xicam.core.workspace import Ensemble
+from xicam.gui.models import EnsembleModel, IntentsModel
+from xicam.gui.widgets.views import DataSelectorView, StackedCanvasView
+from xicam.plugins import GUIPlugin
+
+
+class EnsembleGUIPlugin(GUIPlugin):
```
Comment (ronpandolfi) [inline xicam/gui/plugins/ensembleguiplugin.py:11]: support mixin usage (pass kwargs through to super in init)

File: xicam/gui/plugins/ensembleguiplugin.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        if active_ensemble is not None:
+            ensemble = active_ensemble.data(EnsembleModel.object_role)
+        else:
+            # FIXME who controls creating a new vs appending to existing ensemble?
+            ensemble = Ensemble()
+        ensemble.append_catalog(catalog)
+
+        # TODO: use Dylan's code here instead of default projector
+        def default_projector(catalog):
+            return catalog
```
Comment (ronpandolfi) [inline xicam/gui/plugins/ensembleguiplugin.py]: return empty-list (to hint that the return type should be list of intents); type hint?

File: xicam/gui/models/__init__.py [hunk 10]
```diff
@@ -19,52 +20,116 @@ class EnsembleModel(TreeModel):
...
+        if ensemble_item is not None:
+            end_row = ensemble_item.childCount()
+            # Use begin/end to notify views (e.g. dataselectorview) that it needs to update view after item is inserted
+            self.beginInsertRows(self.index(ensemble_item.row(), 0), end_row, end_row+1)
+            self._create_catalog_item(ensemble_item, catalog, projector)
+            self.endInsertRows()
+        else:
+            self.add_ensemble(ensemble, projector)
+
     def add_ensemble(self, ensemble: Ensemble, projector: Callable):
```
Comment (ronpandolfi) [inline xicam/gui/models/__init__.py]: more specific type hint for projector

File: xicam/gui/plugins/ensembleguiplugin.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        self.intents_model = IntentsModel()
+        self.intents_model.setSourceModel(self.ensemble_model)
+
+        self.data_selector_view = DataSelectorView()
+        self.data_selector_view.setModel(self.ensemble_model)
+
+        self.canvases_view = StackedCanvasView()
+        self.canvases_view.setModel(self.intents_model)
+
+    def appendCatalog(self, catalog: BlueskyRun, projector: Callable = None, **kwargs):
```
Comment (ronpandolfi) [inline xicam/gui/plugins/ensembleguiplugin.py]: better type hint on projector

File: xicam/gui/widgets/imageviewmixins.py [hunk 12]
```diff
@@ -19,11 +19,10 @@
 from functools import partial
 
 from xicam.plugins import manager as pluginmanager, live_plugin
-from .ROI import RectROI
 import inspect
 
 
-# NOTE: PyQt widget mixins have pitfalls; note #2 here: http://trevorius.com/scrapbook/python/pyqt-multiple-inheritance/
+# NOTE: PyQt widget stages have pitfalls; note #2 here: http://trevorius.com/scrapbook/python/pyqt-multiple-inheritance/
```
Comment (ronpandolfi) [inline xicam/gui/widgets/imageviewmixins.py]: revert find/replace in multiple places in this module

File: xicam/gui/widgets/plotwidgetmixins.py [hunk 13]
```diff
@@ -109,13 +110,114 @@ def showLabel(self, item, points):
...
+    def __getattr__(self, attr):
+        # implicitly wrap methods from plotItem
+        if hasattr(self.plot_widget, attr):
+            m = getattr(self.plot_widget, attr)
+            if hasattr(m, '__call__'):
+                return m
+        raise AttributeError(attr)
+
+
+class AutoRotateCheckbox(PlotWidget):
```
Comment (ronpandolfi) [inline xicam/gui/widgets/plotwidgetmixins.py]: maybe we don't need this?

File: xicam/gui/widgets/plotwidgetmixins.py [hunk 14]
```diff
@@ -109,13 +110,114 @@ def showLabel(self, item, points):
...
+        self.inner_layout.addLayout(layout)
+
+    def _offset_toggled(self, enabled):
+        if enabled:
+            self.offset_button.setText("Disable Offset")
+        else:
+            self.offset_button.setText("Enable Offset")
+
+
+class LogButton(PlotWidget):
```
Comment (ronpandolfi) [inline xicam/gui/widgets/plotwidgetmixins.py]: Rename to "XLogButton" (?)

File: xicam/gui/widgets/plotwidgetmixins.py [hunk 15]
```diff
@@ -109,13 +110,114 @@ def showLabel(self, item, points):
...
+        sizePolicy.setHeightForWidth(self.x_log_button.sizePolicy().hasHeightForWidth())
+        self.x_log_button.clicked.connect(self.setXLogMode)
+        self.inner_layout.addWidget(self.x_log_button)
+
+    def setXLogMode(self):
+        log_mode = self.x_log_button.isChecked()
+        self.setLogMode(x=log_mode, y=False)
+
+
+class BetterPlotWidget(AutoRotateCheckbox, OffsetPlots, LogButton):...
```
Comment (ronpandolfi) [inline xicam/gui/widgets/plotwidgetmixins.py]: Annotate example blend as a comment

File: xicam/gui/widgets/views.py [hunk 16]
```diff
@@ -397,10 +399,125 @@ def show_canvases(self, canvases):
             canvas.setVisible(True)
 
 
-class DataSelectorView(QTreeView):
+class MyStyledItemDelegate(QStyledItemDelegate):
```
Comment (ronpandolfi) [inline xicam/gui/widgets/views.py]: remove

File: xicam/gui/windows/mainwindow.py [hunk 17]
```diff
@@ -162,10 +165,14 @@ def open(self, header):
         if self.currentGUIPlugin is None:
             msg.notifyMessage("Please select a gui plugin from the top before trying to open an image.")
             return
+        append = False
```
Comment (ronpandolfi) [inline xicam/gui/windows/mainwindow.py]: hold off on these changes until we can discuss integrating the ensembles/workspace concept into the larger application

File: xicam/plugins/guiplugin.py [hunk 18]
```diff
@@ -19,6 +18,7 @@ class GUIPlugin(PluginType):
 
     is_singleton = True
     needs_qt = True
+    supports_ensembles = False
```
Comment (ronpandolfi) [inline xicam/plugins/guiplugin.py]: also hold off on these for more discussion

**Review by ronpandolfi — CHANGES_REQUESTED** [review]
see comments

File: xicam/gui/models/__init__.py (hunk 1 above)
Comment (birdzix) [reply xicam/gui/models/__init__.py]: this should be addressed now

File: xicam/gui/models/__init__.py (hunk 2 above)
Comment (birdzix) [reply xicam/gui/models/__init__.py]: done

File: xicam/gui/models/__init__.py (hunk 3 above)
Comment (birdzix) [reply xicam/gui/models/__init__.py]: removed prints

File: xicam/gui/models/__init__.py (hunk 4 above)
Comment (birdzix) [reply xicam/gui/models/__init__.py]: implemented using qpalette

File: xicam/gui/models/__init__.py (hunk 6 above)
Comment (birdzix) [reply xicam/gui/models/__init__.py]: :+1:

File: xicam/gui/models/treemodel.py (hunk 7 above)
Comment (birdzix) [reply xicam/gui/models/treemodel.py]: this is fixed

File: xicam/gui/widgets/plotwidgetmixins.py (hunk 14 above)
Comment (birdzix) [reply xicam/gui/widgets/plotwidgetmixins.py]: log buttons have been refactored to include both x and y (you probably will want to be able to toggle log mode for both axes)

File: xicam/gui/widgets/plotwidgetmixins.py (hunk 15 above)
Comment (birdzix) [reply xicam/gui/widgets/plotwidgetmixins.py]: Moved into the __main__ example area for the module

File: xicam/gui/widgets/views.py (hunk 16 above)
Comment (birdzix) [reply xicam/gui/widgets/views.py]: removed

File: xicam/gui/widgets/plotwidgetmixins.py (hunk 13 above)
Comment (birdzix) [reply xicam/gui/widgets/plotwidgetmixins.py]: this has been removed. after further discussion, the intent of this is not to rotate into a "correct" orientation, but to auto-rotate the viewbox according to screen real estate

File: xicam/gui/widgets/imageviewmixins.py (hunk 12 above)
Comment (birdzix) [reply xicam/gui/widgets/imageviewmixins.py]: reverted

File: xicam/gui/models/__init__.py (hunk 10 above)
Comment (birdzix) [reply xicam/gui/models/__init__.py]: :+1:

File: xicam/gui/plugins/ensembleguiplugin.py (hunk 9 above)
Comment (birdzix) [reply xicam/gui/plugins/ensembleguiplugin.py]: :+1:

File: xicam/gui/models/__init__.py (hunk 5 above)
Comment (birdzix) [reply xicam/gui/models/__init__.py]: :+1:

File: xicam/gui/windows/mainwindow.py (hunk 17 above)
Comment (birdzix) [reply xicam/gui/windows/mainwindow.py]: Removed these changes.

File: xicam/plugins/guiplugin.py (hunk 18 above)
Comment (birdzix) [reply xicam/plugins/guiplugin.py]: Removed these changes to GUIPlugin, should be back to original state in `main`. I kept SAXS/XPCS inheriting the `EnsembleGUIPlugin`, I think that should be okay?

**Review by ronpandolfi — APPROVED** [review]

**Comment by ronpandolfi** [comment]
TODO post-merge: add tests for mixins

**Pull request merged**